pkg/sidecar-controller/groupsnapshot_helper.go

```diff
@@ -159,12 +167,477 @@ func (ctrl *csiSnapshotSideCarController) syncGroupSnapshotContent(content *crdv
+		contentClone.Status = newStatus
+		newContent, err := ctrl.clientset.GroupsnapshotV1alpha1().VolumeGroupSnapshotContents().UpdateStatus(context.TODO(), contentClone, metav1.UpdateOptions{})
+		if err != nil {
+			return contentObj, newControllerUpdateError(content.Name, err.Error())
+		}
+		return newContent, nil
+	}
+
+	return contentObj, nil
+}
+
+// updateContentStatusWithEvent saves new content.Status to API server and emits
+// given event on the content. It saves the status and emits the event only when
+// the status has actually changed from the version saved in API server.
+// Parameters:
+//
+// * content - content to update
+// * eventtype, reason, message - event to send, see EventRecorder.Event()
+func (ctrl *csiSnapshotSideCarController) updateGroupSnapshotContentErrorStatusWithEvent(content *crdv1alpha1.VolumeGroupSnapshotContent, eventtype, reason, message string) error {
+	klog.V(5).Infof("updateContentStatusWithEvent[%s]", content.Name)
+
+	if content.Status != nil && content.Status.Error != nil && *content.Status.Error.Message == message {
+		klog.V(4).Infof("updateContentStatusWithEvent[%s]: the same error %v is already set", content.Name, content.Status.Error)
+		return nil
+	}
+
+	var patches []utils.PatchOp
+	ready := false
+	contentStatusError := &crdv1.VolumeSnapshotError{
+		Time: &metav1.Time{
+			Time: time.Now(),
+		},
+		Message: &message,
+	}
+	if content.Status == nil {
+		// Initialize status if nil
+		patches = append(patches, utils.PatchOp{
+			Op:   "replace",
+			Path: "/status",
+			Value: &crdv1alpha1.VolumeGroupSnapshotContentStatus{
+				ReadyToUse: &ready,
+				Error:      contentStatusError,
+			},
+		})
+	} else {
+		// Patch status if non-nil
+		patches = append(patches, utils.PatchOp{
+			Op:    "replace",
+			Path:  "/status/error",
+			Value: contentStatusError,
+		})
+		patches = append(patches, utils.PatchOp{
+			Op:    "replace",
+			Path:  "/status/readyToUse",
+			Value: &ready,
+		})
+
+	}
+
+	newContent, err := utils.PatchVolumeGroupSnapshotContent(content, patches, ctrl.clientset, "status")
+
+	// Emit the event even if the status update fails so that user can see the error
+	ctrl.eventRecorder.Event(newContent, eventtype, reason, message)
+
+	if err != nil {
+		klog.V(4).Infof("updating VolumeGroupSnapshotContent[%s] error status failed %v", content.Name, err)
+		return err
+	}
+
+	_, err = ctrl.storeGroupSnapshotContentUpdate(newContent)
+	if err != nil {
+		klog.V(4).Infof("updating VolumeGroupSnapshotContent[%s] error status: cannot update internal cache %v", content.Name, err)
+		return err
+	}
+
+	return nil
+}
+
+// GetSnapshotNameForVolumeGroupSnapshotContent returns a unique snapshot name for a VolumeGroupSnapshotContent.
+func GetSnapshotNameForVolumeGroupSnapshotContent(content *crdv1alpha1.VolumeGroupSnapshotContent) string {
+	return fmt.Sprintf("groupsnapshot-%x-%d", sha256.Sum256([]byte(content.UID)), time.Duration(time.Now().UnixNano())/time.Millisecond)
```

Comment:
snapshot-<hash of VolumeGroupSnapshot UUID+PVC UUID>-timestamp

Reply:
Done - https://github.com/kubernetes-csi/external-snapshotter/pull/826/commits/c03573d80cf9a8b98917076d2a3af423b6c9efc0